File: app/src/main/java/com/elementary/tasks/core/controller/TimerEvent.java
package com.elementary.tasks.core.controller;

import android.content.Context;
import android.util.Log;

import com.elementary.tasks.core.services.AlarmReceiver;
import com.elementary.tasks.core.services.DelayReceiver;
import com.elementary.tasks.core.utils.TimeCount;
import com.elementary.tasks.core.utils.TimeUtil;
import com.elementary.tasks.reminder.models.Reminder;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

class TimerEvent extends RepeatableEventManager {

    private static final String TAG = "TimerEvent";

    TimerEvent(Reminder reminder, Context context) {
        super(reminder, context);
    }

    @Override
    public boolean start() {
        if (TimeCount.isCurrent(mReminder.getEventTime())) {
            mReminder.setActive(true);
            super.save();
            new AlarmReceiver().enableReminder(mContext, mReminder.getUuId());
            super.export();
            return true;
        }
        return false;
    }

    @Override
    public boolean stop() {
        return super.stop();
    }

    @Override
    public boolean pause() {
        new AlarmReceiver().cancelAlarm(mContext, mReminder.getUniqueId());
        return true;
    }

    @Override
    public boolean skip() {
        return false;
    }

    @Override
    public boolean resume() {
        new AlarmReceiver().enableReminder(mContext, mReminder.getUuId());
        return true;
    }

    @Override
    public boolean next() {
        if (!isRepeatable()) {
            return stop();
        } else {
            long time = calculateTime(false);
            while (time < System.currentTimeMillis()) {
                mReminder.setEventTime(TimeUtil.getGmtFromDateTime(time));
                time = calculateTime(false);
            }
            Log.d(TAG, "next: " + TimeUtil.getFullDateTime(time, true, true));
            mReminder.setEventTime(TimeUtil.getGmtFromDateTime(time));
            mReminder.setEventCount(mReminder.getEventCount() + 1);
            return start();
        }
    }

    @Override
    public boolean onOff() {
        if (isActive()) {
            return stop();
        } else {
            long time = calculateTime(true);
            while (time < System.currentTimeMillis()) {
                mReminder.setEventTime(TimeUtil.getGmtFromDateTime(time));
                time = calculateTime(true);
            }
            mReminder.setEventTime(TimeUtil.getGmtFromDateTime(time));
            mReminder.setEventCount(0);
            return start();
        }
    }

    @Override
    public boolean isActive() {
        return mReminder.isActive();
    }

    @Override
    public boolean canSkip() {
        return isRepeatable() && (mReminder.getRepeatLimit() == -1 || mReminder.getRepeatLimit() - mReminder.getEventCount() - 1 > 0);
    }

    @Override
    public boolean isRepeatable() {
        return mReminder.getRepeatInterval() > 0;
    }

    @Override
    public void setDelay(int delay) {
        if (delay == 0) {
            next();
            return;
        }
        mReminder.setDelay(delay);
        super.save();
        new DelayReceiver().setAlarm(mContext, mReminder.getUniqueId(), delay, mReminder.getUuId());
    }

    @Override
    public long calculateTime(boolean isNew) {
        return TimeCount.getInstance(mContext).generateNextTimer(mReminder, isNew);
    }
}

File: app/src/main/java/com/elementary/tasks/core/controller/TimerEvent.kt
package com.elementary.tasks.core.controller

import com.elementary.tasks.core.appwidgets.UpdatesHelper
import com.elementary.tasks.core.services.JobScheduler
import com.elementary.tasks.core.utils.GoogleCalendarUtils
import com.elementary.tasks.core.utils.Notifier
import com.elementary.tasks.core.utils.datetime.DateTimeManager
import com.elementary.tasks.core.utils.params.Prefs
import com.github.naz013.domain.Reminder
import com.github.naz013.feature.common.android.TextProvider
import com.github.naz013.logging.Logger
import com.github.naz013.repository.GoogleTaskRepository
import com.github.naz013.repository.ReminderRepository
import org.threeten.bp.LocalDateTime

class TimerEvent(
  reminder: Reminder,
  reminderRepository: ReminderRepository,
  prefs: Prefs,
  googleCalendarUtils: GoogleCalendarUtils,
  notifier: Notifier,
  jobScheduler: JobScheduler,
  updatesHelper: UpdatesHelper,
  textProvider: TextProvider,
  private val dateTimeManager: DateTimeManager,
  googleTaskRepository: GoogleTaskRepository
) : RepeatableEventManager(
  reminder,
  reminderRepository,
  prefs,
  googleCalendarUtils,
  notifier,
  jobScheduler,
  updatesHelper,
  textProvider,
  dateTimeManager,
  googleTaskRepository
) {

  override val isActive: Boolean
    get() = reminder.isActive

  override fun justStart() {
    reminder.isActive = true
    reminder.isRemoved = false
    super.save()
    super.enableReminder()
    super.export()
  }

  override fun enable(): Boolean {
    if (dateTimeManager.isCurrent(reminder.eventTime)) {
      reminder.isActive = true
      reminder.isRemoved = false
      super.save()
      super.enableReminder()
      super.export()
    }
    return true
  }

  override fun skip(): Boolean {
    reminder.delay = 0
    if (canSkip()) {
      val time = calculateTime(false)
      reminder.eventTime = dateTimeManager.getGmtFromDateTime(time)
      enable()
      return true
    }
    return false
  }

  override fun next(): Boolean {
    reminder.delay = 0
    return if (canSkip()) {
      var time = calculateTime(false)
      while (!dateTimeManager.isCurrent(time)) {
        reminder.eventTime = dateTimeManager.getGmtFromDateTime(time)
        time = calculateTime(false)
      }
      Logger.d("next: ${dateTimeManager.logDateTime(time)}")
      reminder.eventTime = dateTimeManager.getGmtFromDateTime(time)
      reminder.eventCount += 1
      enable()
    } else {
      disable()
    }
  }

  override fun onOff(): Boolean {
    return if (isActive) {
      disable()
    } else {
      var time = calculateTime(true)
      while (!dateTimeManager.isCurrent(time)) {
        reminder.eventTime = dateTimeManager.getGmtFromDateTime(time)
        time = calculateTime(true)
      }
      reminder.eventTime = dateTimeManager.getGmtFromDateTime(time)
      reminder.eventCount = 0
      enable()
    }
  }

  override fun canSkip(): Boolean {
    return reminder.isRepeating() && (!reminder.isLimited() || !reminder.isLimitExceed())
  }

  override fun setDelay(delay: Int) {
    if (delay == 0) {
      next()
      return
    }
    reminder.delay = delay
    super.save()
    super.setDelay(delay)
  }

  override fun calculateTime(isNew: Boolean): LocalDateTime {
    return dateTimeManager.generateNextTimer(reminder, isNew)
  }
}
